samza-kafka/src/main/java/org/apache/samza/system/kafka/KafkaSystemAdmin.java

```diff
@@ -0,0 +1,662 @@
+  protected static final long DEFAULT_EXPONENTIAL_SLEEP_INITIAL_DELAY_MS = 500;
+  protected static final long DEFAULT_EXPONENTIAL_SLEEP_MAX_DELAY_MS = 10000;
+  protected static final int MAX_RETRIES_ON_EXCEPTION = 5;
+  public static final int DEFAULT_REPL_FACTOR = 2;
+
+  // used in TestRepartitionJoinWindowApp TODO - remove SAMZA-1945
+  @VisibleForTesting
+  public static volatile boolean deleteMessageCalled = false;
+
+  protected final String systemName;
+  protected final Consumer metadataConsumer;
+
+  // get ZkUtils object to connect to Kafka's ZK.
+  private final Supplier<ZkUtils> getZkConnection;
+
+  // Custom properties to create a new coordinator stream.
+  private final Properties coordinatorStreamProperties;
+
+  // Replication factor for a new coordinator stream.
+  private final int coordinatorStreamReplicationFactor;
+
+  // Replication factor and kafka properties for changelog topic creation
+  private final Map<String, ChangelogInfo> changelogTopicMetaInformation;
+
+  // Kafka properties for intermediate topics creation
+  private final Map<String, Properties> intermediateStreamProperties;
+
+  // adminClient is required for deleteCommittedMessages operation
+  private final AdminClient adminClient;
+
+  // used for intermediate streams
+  private final boolean deleteCommittedMessages;
+
+  private final AtomicBoolean stopped = new AtomicBoolean(false);
+
+  public KafkaSystemAdmin(String systemName, Config config, Consumer metadataConsumer) {
+    this.systemName = systemName;
+
+    if (metadataConsumer == null) {
+      throw new SamzaException("Cannot construct KafkaSystemAdmin for system " + systemName + " with null metadataConsumer");
+    }
+    this.metadataConsumer = metadataConsumer;
+
+    // populate brokerList from either consumer or producer configs
+    Properties props = new Properties();
+    String brokerList =
+        config.get(String.format(CONSUMER_CONFIGS_CONFIG_KEY, systemName, ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG));
+    if (brokerList == null) {
+      brokerList = config.get(String.format(PRODUCER_CONFIGS_CONFIG_KEY, systemName, ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG));
+    }
+    if (brokerList == null) {
+      throw new SamzaException(
+          ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG + " is required for systemAdmin for system " + systemName);
+    }
+    props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerList);
+
+    // kafka.admin.AdminUtils requires zkConnect
+    // this will change after we move to the new org.apache..AdminClient
+    String zkConnect = config.get(String.format(CONSUMER_CONFIGS_CONFIG_KEY, systemName, ZOOKEEPER_CONNECT));
+    if (StringUtils.isBlank(zkConnect)) {
+      throw new SamzaException("Missing zookeeper.connect config for admin for system " + systemName);
+    }
+    props.put(ZOOKEEPER_CONNECT, zkConnect);
+
+    adminClient = AdminClient.create(props);
+
+    getZkConnection = () -> {
```

Comment:
I think i'll be cleaner to inline this or make this a private helper method (`createZkUtils`)